enhancements/machine-api/out-of-tree-provider-support.md

```diff
@@ -0,0 +1,553 @@
+
+#### Resource changes
+
+#### Kubelet
+
+In OpenShift, `machine-config-operator` manages `kubelet` configuration.
+Based on the [Infrastructure value](https://github.com/openshift/api/blob/master/config/v1/types_infrastructure.go) MCO currently sets the correct `cloud-config` value.
+To support external providers, MCO should not set the `cloud-config` value anymore.
+
+Additionally, it must set the value of the `cloud-provider` [flag](https://github.com/openshift/machine-config-operator/blob/2a51e49b0835bbd3ac60baa685299e86777b064f/templates/worker/01-worker-kubelet/_base/units/kubelet.service.yaml#L32) to `external`.
+This informs `kubelet` that an out-of-tree provider will finish the `Node` initialization process.
+
+While `kubelet` itself does not directly contact or interact with the `CCM`, formation of a functional `Node` in Kubernetes does rely on the `CCM`.
+With in-tree providers, `kubelet` looks up Node IP Addresses, Instance IDs and other metadata directly from the cloud provider and places this information onto the `Node` object.
+
+When using out-of-tree, the `kubelet` taints the `Node` with `node.cloudprovider.kubernetes.io/uninitialized: true` and does not look up any information.
+It is then the responsibility of the `CCM` to look up the Node IP Addresses, Instance IDs and other metadata and labels, place the information onto the `Node` object and then remove the taint from the `Node`.
+
+The information previously set by `kubelet`, now `CCM`, is required to make a `Node` schedulable.
+Without this information, the `scheduler` cannot schedule `Pods` that have any sort of affinity that relies on `Node` metadata (eg failure domain).
+
+#### Kube Controller Manager and Kube API Server
+
+In OpenShift, `kube-controller-manager` and `kube-apiserver` are managed by `cluster-kube-controller-manager-operator` and `cluster-kube-apiserver-operator` respectively.
+When switching to out-of-tree providers, the `cloud-provider` option [should be removed](https://kubernetes.io/docs/tasks/administer-cluster/running-cloud-controller/#running-cloud-controller-manager) from their deployments.
+
+##### Kube Controller Manager changes to prolong support for in-tree storage
+
+In the first phase of the transition all cloud providers will still rely on the in-tree storage. `--external-cloud-volume-plugin=<some-provider>` set on `KCM` pod will preserve the in-tree storage controller loops.
+
+At the later phases, CSI support and CSI migration will be responsible for removing this flag from the `KCM`.
+
+With the `--external-cloud-volume-plugin` flag set, existing designs for this transition are being preserved. The storage will temporarily become a lone in-tree controller depending on the cloud, and could be excluded by setting this flag off. This means that `kcm` will still initialize the `cloudprovider.Interface`, but this will only be used by storage controllers.
```

Comment:
i found this paragraph slightly confusing. i think you are saying the following things:

1. the `--external-cloud-volume-plugin` flag will allow us to preserve existing, in-tree, functionality for storage
2. not setting this flag will preserve the in-tree usage

the confusing part to me is this:
> With the `--external-cloud-volume-plugin` flag set, existing designs for this transition are being preserved.

i think the message here is that by having this flag we can preserve the existing in-tree design (eg not using the flag), while enabling the out-of-tree design (eg by setting the flag). is that accurate?

Reply:
It is slightly different then that. `--external-cloud-volume-plugin=aws` will preserve in-tree PV and PVC controllers hosted in kube-controller-manager, while the `--cloud-provider=external` set *at the same time* will effectively disable some controllers (there is a map of them, don't give precise naming for that reason - [here](https://github.com/kubernetes/kubernetes/blob/master/cmd/kube-controller-manager/app/controllermanager.go#L432-L437) )yet others, like https://github.com/kubernetes/kubernetes/blob/master/cmd/kube-controller-manager/app/controllermanager.go#L439 will still use initialized cloud interface and be enabled. Volumes could be attached and detached. 

Missing piece is CSI, which will allow full operation on volumes, like creation provisioning for PVC etc. This flag is set only on KCM, which leads to `kubelet` lacking ~any~ kind of initialization for cloud. Therefor all storage related parts located in `kubelet` will remain non-functional. Until CSI is enabled

Reply:
ok, thanks for that detailed explanation. i think this section should be cleaned up to reflect what you said here, it makes it much clearer imo.